### PR TITLE
Add system font change listener for windows

### PR DESCRIPTION
**Pull request by chunhtai**
This pr add listener in windows embedding and use the new embedding api to notify system font changes
This requires framework https://github.com/flutter/flutter/pull/38930
This also requires example app change https://github.com/google/flutter-desktop-embedding/pull/563

https://github.com/flutter/flutter/issues/38556

### CODE REVIEW

**Comment by chunhtai** [comment]
@googlebot I fixed it

File: shell/platform/embedder/tests/embedder_unittests.cc [hunk 1]
```diff
@@ -228,6 +228,16 @@ TEST(EmbedderTestNoFixture, CanGetCurrentTimeInNanoseconds) {
   ASSERT_LT((point2 - point1), fml::TimeDelta::FromMilliseconds(1));
 }
 
+TEST_F(EmbedderTest, CanReloadSystemFonts) {
```
Comment (chunhtai) [inline shell/platform/embedder/tests/embedder_unittests.cc]: This test make sure the hook from embedding -> embedding_engine -> shell works. It does not test for functionality.

File: shell/common/shell_unittests.cc [hunk 2]
```diff
@@ -519,6 +519,30 @@ TEST_F(ShellTest, ReportTimingsIsCalledImmediatelyAfterTheFirstFrame) {
   ASSERT_EQ(timestamps.size(), FrameTiming::kCount);
 }
 
+TEST_F(ShellTest, ReloadSystemFonts) {
```
Comment (chunhtai) [inline shell/common/shell_unittests.cc]: This test the actual functionality

File: shell/platform/windows/win32_flutter_window.cc [hunk 3]
```diff
@@ -147,6 +151,22 @@ void Win32FlutterWindow::OnClose() {
   messageloop_running_ = false;
 }
 
+void Win32FlutterWindow::OnFontChange() {
```
Comment (chunhtai) [inline shell/platform/windows/win32_flutter_window.cc]: @kpozin  Here is the current design for system font api.
When system font changes:
1. Use FlutterEngineReloadSystemFonts in embedder.h to reload system font and clear caches
2. send system channel message {"type": "fontsChange"} to "flutter/system"  flutter framework.

Noted that you will probably need to implement fuchsia version of loading default fonts.
see:   third_party/txt/src/txt/platform_windows.cc
FlutterEngineReloadSystemFonts relys on this function to retrieve the latest system fonts. If there is any caching happen under this layer, the FlutterEngineReloadSystemFonts will still load the old value.

Let me know if this looks ok to you.

Comment (kpozin) [reply shell/platform/windows/win32_flutter_window.cc]: Yes, this seems fine to me.

Comment (chunhtai) [reply shell/platform/windows/win32_flutter_window.cc]: Thanks for confirming, I will let you know if anything change.

File: shell/common/shell_unittests.cc [hunk 4]
```diff
@@ -519,6 +519,32 @@ TEST_F(ShellTest, ReportTimingsIsCalledImmediatelyAfterTheFirstFrame) {
...
+  auto task_runner = fml::MessageLoop::GetCurrent().GetTaskRunner();
+  TaskRunners task_runners("test", task_runner, task_runner, task_runner,
+                           task_runner);
+  auto shell = CreateShell(std::move(settings), std::move(task_runners));
+
+  auto fontCollection = GetFontCollection(shell.get());
+  std::vector<std::string> families(1, "Robotofake");
+  auto font =
+      fontCollection->GetMinikinFontCollectionForFamilies(families, "en");
+  ASSERT_EQ(font->getId(), (unsigned int)0);
```
Comment (stuartmorgan) [inline shell/common/shell_unittests.cc]: C-style casts are forbidden; use C++ casts (static_cast here)

File: shell/platform/windows/client_wrapper/flutter_view_controller.cc [hunk 5]
```diff
@@ -58,5 +58,8 @@ FlutterDesktopPluginRegistrarRef FlutterViewController::GetRegistrarForPlugin(
   }
   return FlutterDesktopGetPluginRegistrar(controller_, plugin_name.c_str());
 }
+void FlutterViewController::OnFontChange() {
```
Comment (stuartmorgan) [inline shell/platform/windows/client_wrapper/flutter_view_controller.cc]: Blank line above this.

File: shell/platform/windows/client_wrapper/include/flutter/flutter_view_controller.h [hunk 6]
```diff
@@ -61,6 +61,8 @@ class FlutterViewController {
   // loop.
   void ProcessMessages();
 
+  void OnFontChange();
```
Comment (stuartmorgan) [inline shell/platform/windows/client_wrapper/include/flutter/flutter_view_controller.h]: Needs a declaration comment.

It should also get a TODO to move out of the view controller class once there is an engine class. You can reference #38600

File: shell/platform/windows/public/flutter_windows.h [hunk 7]
```diff
@@ -69,6 +69,10 @@ FlutterDesktopGetHWND(FlutterDesktopViewControllerRef controller);
 // loop.
 FLUTTER_EXPORT void FlutterDesktopProcessMessages();
 
+// Invokes OnFontChange callback in flutter windows view.
+FLUTTER_EXPORT void FlutterOnFontChange(
```
Comment (stuartmorgan) [inline shell/platform/windows/public/flutter_windows.h]: Everything in this API surface should use `FlutterDesktop` as a prefix.

File: shell/platform/windows/win32_flutter_window.cc [hunk 8]
```diff
@@ -1,6 +1,10 @@
 #include "flutter/shell/platform/windows/win32_flutter_window.h"
-
 #include <chrono>
+#include "flutter/shell/platform/common/cpp/client_wrapper/include/flutter/json_message_codec.h"
```
Comment (stuartmorgan) [inline shell/platform/windows/win32_flutter_window.cc]: Blank line above this.

File: shell/platform/windows/win32_flutter_window.cc [hunk 9]
```diff
@@ -1,6 +1,10 @@
...
+
+static constexpr char kSystemChannelName[] = "flutter/system";
```
Comment (stuartmorgan) [inline shell/platform/windows/win32_flutter_window.cc]: win32_flutter_window doesn't seem like the right place for the flutter/system channel to live. Please set up a new handler for now, following the model of platform_handler.{h,cc}. Later we may decide to fold it into another class, but that will be easier if it's not embedded into this one to start with.

File: shell/platform/embedder/embedder.h [hunk 10]
```diff
@@ -1192,6 +1192,17 @@ FlutterEngineResult FlutterEngineOnVsync(FLUTTER_API_SYMBOL(FlutterEngine)
...
 
+//------------------------------------------------------------------------------
+/// @brief      Reloads the system fonts in engine.
+///
+/// @param[in]  engine.                  A running engine instance.
+///
+/// @return     The result of the call.
+///
+FLUTTER_EXPORT
+FlutterEngineResult FlutterEngineReloadSystemFonts(
```
Comment (stuartmorgan) [inline shell/platform/embedder/embedder.h]: What happens if someone calls this but doesn't send the platform event? It seems like if both are required that should be documented.

Although it would be even better if that were handled internally. Couldn't the implementation send a system channel message? `shell/common/shell.cc` uses the `flutter/skia` channel for instance.

File: shell/platform/windows/win32_window.cc [hunk 11]
```diff
@@ -129,7 +129,9 @@ Win32Window::MessageHandler(HWND hwnd,
         current_height_ = height;
         window->HandleResize(width, height);
         break;
-
+      case WM_FONTCHANGE:
```
Comment (stuartmorgan) [inline shell/platform/windows/win32_window.cc]: Does this ever happen? IIRC the docs said that this is sent to top-level windows, which this isn't.

If we can get this here, then I'm not sure why we need runner-level changes and embedding API changes.

File: third_party/txt/src/txt/platform_windows.cc [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "third_party/skia/include/ports/SkTypeface_win.h"
+#include "txt/platform.h"
+
+namespace txt {
+
+std::string GetDefaultFontFamily() {
+  return "Arial";
```
Comment (stuartmorgan) [inline third_party/txt/src/txt/platform_windows.cc]: This change seems unrelated. Is this improving font fallback handling? If so, it would be great to split this out into its own change as part of #39915

Also, I'm surprised to see a file being added to third_party. What is the source for `txt`, and why isn't this an upstream change?

Comment (chunhtai) [reply third_party/txt/src/txt/platform_windows.cc]: This is essential for reloading system fonts. The original GetDefaultFontFamily in third_party/txt/src/txt/platform.cc uses SkFontMgr::RefDefault() which will cache the result and there is no way to clear it. When a new system font is installed, it will not load the latest system fonts but return previous cached. We have to implement the windows version of it.

There is no upstream for txt. The txt's source lives in here. One thing that worries me is that this might be migrated to skia library at some point @jason-simmons Do you think it is ok to change it now?

File: shell/platform/windows/win32_window.cc (hunk 11 above)
Comment (chunhtai) [reply shell/platform/windows/win32_window.cc]: I should be able to remove this. The only way this might be needed is application sends targeted windows message, but that probably won't happen in real life

File: shell/platform/embedder/embedder.h (hunk 10 above)
Comment (chunhtai) [reply shell/platform/embedder/embedder.h]: I can't think of a use case at the moment where someone only reload the system fonts but don't want to rebuild flutter app .
I should be able to combine them

File: shell/platform/windows/win32_window.cc (hunk 11 above)
Comment (stuartmorgan) [reply shell/platform/windows/win32_window.cc]: Per the latest discussion in the other PR, I think what we actually want to do is keep this, remove the embedding-level APIs, and forward the WM_FONTCHANGE message down in the top-level window implementation.

File: third_party/txt/src/txt/platform_windows.cc (hunk 12 above)
Comment (jason-simmons) [reply third_party/txt/src/txt/platform_windows.cc]: Yes - go ahead and modify this in libtxt.  if we later migrate to Skia's text shaper then we will need to port this feature over to the integration with Skia.

File: shell/common/shell.cc [hunk 13]
```diff
@@ -1357,4 +1362,36 @@ fml::Status Shell::WaitForFirstFrame(fml::TimeDelta timeout) {
...
+  document.Parse("{}");
+  auto& allocator = document.GetAllocator();
+  rapidjson::Value message_value;
+  message_value.SetString(kFontChange, allocator);
+  document.AddMember(kTypeKey, message_value, allocator);
+
+  rapidjson::StringBuffer buffer;
+  rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
+  document.Accept(writer);
+  std::string message(buffer.GetString(), buffer.GetSize());
```
Comment (stuartmorgan) [inline shell/common/shell.cc]: IIRC you only need `GetString()`, since we should be building rapidjson with std::string support. Am I misremembering?

File: shell/platform/windows/client_wrapper/flutter_view_controller.cc [hunk 14]
```diff
@@ -58,5 +58,4 @@ FlutterDesktopPluginRegistrarRef FlutterViewController::GetRegistrarForPlugin(
   }
   return FlutterDesktopGetPluginRegistrar(controller_, plugin_name.c_str());
 }
-
```
Comment (stuartmorgan) [inline shell/platform/windows/client_wrapper/flutter_view_controller.cc]: Revert please, since you're no longer touching this file otherwise.

File: shell/platform/windows/client_wrapper/testing/stub_flutter_windows_api.cc [hunk 15]
```diff
@@ -70,6 +70,8 @@ void FlutterDesktopProcessMessages() {
   }
 }
 
+void FlutterOnFontChange(FlutterDesktopViewControllerRef controller) {}
```
Comment (stuartmorgan) [inline shell/platform/windows/client_wrapper/testing/stub_flutter_windows_api.cc]: Remove these changes.

File: shell/platform/windows/win32_flutter_window.cc (hunk 9 above)
Comment (stuartmorgan) [reply shell/platform/windows/win32_flutter_window.cc]: This is all cruft now.

**Comment by chunhtai** [comment]
@stuartmorgan  this is ready for re-review

File: shell/platform/windows/win32_flutter_window.cc (hunk 3 above)
Comment (chunhtai) [reply shell/platform/windows/win32_flutter_window.cc]: @kpozin a little update
you will only need to call this embedder api https://github.com/flutter/engine/blob/aadd5a346bb5fcf970a90e3c98d55818b4734e48/shell/platform/embedder/embedder.h#L1211
I make it so it will send the system message as part of the api.
However, you will still need to implement fuchsia version of loading default fonts.
https://github.com/flutter/engine/blob/master/third_party/txt/src/txt/platform_fuchsia.cc
